pallets/msa/src/lib.rs

```diff
@@ -905,6 +924,99 @@ pub mod pallet {
+		/// * [`Error::NoKeyExists`] - the public key supplied in `msa_owner_public_key` is not a registered control key of any MSA.
+		/// * [`Error::NotMsaOwner`] - the MSA ID in `authorization_payload` does not match the MSA ID of the `msa_owner_public_key`.
+		/// * [`Error::ProofHasExpired`] - the current block is less than the `expired` block number set in `AddKeyData`.
+		/// * [`Error::ProofNotYetValid`] - the `expired` block number set in `AddKeyData` is greater than the current block number plus mortality_block_limit().
+		/// * [`Error::SignatureAlreadySubmitted`] - signature has already been used.
+		/// * [`Error::InsufficientBalanceToWithdraw`] - the MSA account has not balance to withdraw
+		/// * [`Error::InsufficientBalanceToFundDestination`] - the transfer would result in an account that would be reaped due to existential balance requirements
+		/// * [`Error::UnexpectedTokenTransferError`] - the balances pallet threw an error during the attempted transfer
+		///
+		#[pallet::call_index(14)]
+		#[pallet::weight(T::WeightInfo::withdraw_tokens())]
+		pub fn withdraw_tokens(
+			origin: OriginFor<T>,
+			msa_owner_public_key: T::AccountId,
+			msa_owner_proof: MultiSignature,
+			authorization_payload: AddKeyData<T>,
+		) -> DispatchResult {
+			let public_key = ensure_signed(origin)?;
+
+			ensure!(public_key == authorization_payload.new_public_key, Error::<T>::NotKeyOwner);
+
+			ensure!(
+				Self::verify_signature(
+					&msa_owner_proof,
+					&msa_owner_public_key,
+					&authorization_payload
+				),
+				Error::<T>::MsaOwnershipInvalidSignature
+			);
+
+			Self::register_signature(&msa_owner_proof, authorization_payload.expiration)?;
+
+			let msa_id = authorization_payload.msa_id;
+
+			Self::ensure_msa_owner(&msa_owner_public_key, msa_id)?;
+
+			// - Get account address for MSA
+			let msa_address = Self::msa_id_to_eth_address(msa_id);
+
+			// - Convert to AccountId
+			let mut bytes = &EthereumAddressMapper::to_bytes32(&msa_address.0)[..];
+			let msa_account_id = T::AccountId::decode(&mut bytes).unwrap();
+
+			// - Check that the MSA has a balance to withdraw
+			let msa_balance = T::Currency::reducible_balance(
+				&msa_account_id,
+				Preservation::Expendable,
+				Fortitude::Polite,
+			);
+			ensure!(msa_balance > Zero::zero(), Error::<T>::InsufficientBalanceToWithdraw);
+
+			// - TODO: Phase 2 (free transaction w/fee paid from proceeds): Check that MSA balance is > fee
+
+			let ed = T::Currency::minimum_balance();
+			// Make sure either:
+			// - Balance to be transferred > existential deposit, OR
+			// - Destination account balance  + ammount to be transferred > existential deposit
+			let amount_alone_is_insufficient = msa_balance < ed;
+			let dest_will_be_reaped = if amount_alone_is_insufficient {
+				let dest_balance = T::Currency::total_balance(&public_key);
+				(dest_balance + msa_balance) < ed
+			} else {
+				amount_alone_is_insufficient
+			};
```

Comment:
these conditions are written in a confusing way

```
let y = if x {
    z
} else {
   x
}
```

Reply:
I'm wondering if these checks are overkill... I can't even engineer a test for them, because it's impossible to fund an account with < ED. Is there any case where an address could be left with dust where it *wouldn't* be reaped? Is there an edge case where an account is scheduled to be reaped, but hasn't yet? I know there is some special handling if free balance < ED but there is locked balance as well, but since these accounts have no control keys, it would be impossible to lock a balance in them.

Reply:
I think I've seen some code that indicates that there might be some edge cases that the account might have less than ED amount but that might be legacy stuff.
@enddynayn  do you have anything to add in here?

Reply:
After some discussion, I'm going to remove the ED check for now. It saves us 1 read (destination balance), and we can't envision any edge cases where an MSA account could possibly have < ED in free balance.